src/Language/PureScript/TypeChecker.hs

```diff
@@ -434,31 +434,98 @@ typeCheckModule (Module ss coms mn decls (Just exps)) =
+
+      superClassesFor qname =
+        fromMaybe S.empty (M.lookup qname classesToSuperClasses)
+
     for_ exps $ \e -> do
       checkTypesAreExported e
       checkClassMembersAreExported e
       checkClassesAreExported e
+      checkSuperClassExport superClassesFor transitiveSuperClassesFor moduleClassDeclarations e
     return $ Module ss coms mn decls' (Just exps)
   where
+  qualify' :: a -> Qualified a
+  qualify' = Qualified (Just mn)
+
+  moduleClassExports :: S.Set (Qualified (ProperName 'ClassName))
+  moduleClassExports = S.fromList $ mapMaybe (\x -> case x of
+     TypeClassRef _ name -> Just (qualify' name)
+     _ -> Nothing) exps
+
+  untilSame :: Eq a => (a -> a) -> a -> a
+  untilSame f a = let a' = f a in if a == a' then a else untilSame f a'
 
   checkMemberExport :: (Type -> [DeclarationRef]) -> DeclarationRef -> m ()
   checkMemberExport extract dr@(TypeRef _ name dctors) = do
     env <- getEnv
-    case M.lookup (Qualified (Just mn) name) (typeSynonyms env) of
-      Nothing -> return ()
-      Just (_, ty) -> checkExport dr extract ty
-    case dctors of
-      Nothing -> return ()
-      Just dctors' -> for_ dctors' $ \dctor ->
-        case M.lookup (Qualified (Just mn) dctor) (dataConstructors env) of
-          Nothing -> return ()
-          Just (_, _, ty, _) -> checkExport dr extract ty
-    return ()
+    for_ (M.lookup (qualify' name) (typeSynonyms env)) $ \(_, ty) ->
+      checkExport dr extract ty
+    for_ dctors $ \dctors' ->
+      for_ dctors' $ \dctor ->
+        for_ (M.lookup (qualify' dctor) (dataConstructors env)) $ \(_, _, ty, _) ->
+          checkExport dr extract ty
   checkMemberExport extract dr@(ValueRef _ name) = do
-    ty <- lookupVariable (Qualified (Just mn) name)
+    ty <- lookupVariable (qualify' name)
     checkExport dr extract ty
   checkMemberExport _ _ = return ()
 
+  checkSuperClassExport
+    :: (Qualified (ProperName 'ClassName) -> S.Set (Qualified (ProperName 'ClassName)))
+    -> (Qualified (ProperName 'ClassName) -> S.Set (Qualified (ProperName 'ClassName)))
+    -> M.Map (Qualified (ProperName 'ClassName)) DeclarationRef
+    -> DeclarationRef
+    -> m ()
+  checkSuperClassExport superClassesFor transitiveSuperClassesFor classMap dr@(TypeClassRef _ className) = do
+    let superClasses = superClassesFor (qualify' className)
+        -- thanks to laziness, the computation of the transitive
+        -- superclasses defined in-module will only occur if we actually
+        -- throw the error. Constructing the full set of transitive
+        -- superclasses is likely to be costly for every single term.
+        transitiveSuperClasses = transitiveSuperClassesFor (qualify' className)
+        unexported = S.difference superClasses moduleClassExports
+    unless (null unexported)
+      . throwError . errorMessage' (declRefSourceSpan dr)
+      . TransitiveExportError dr
+      $ mapMaybe (\n -> M.lookup n classMap) (toList transitiveSuperClasses)
```

Comment:
I think we can get rid of the `classMap` argument here; the `DeclarationRef` values in the second argument to `TransitiveExportError` aren't intended to have real source spans (since they are things which are supposed to be in the exports list but aren't), so I think we can just reuse the one from `dr`. That is what the other `TransitiveExportError` checks seem to be doing, too.

Do you think it would work to use `map (TypeClassRef (declRefSourceSpan dr))` on this line instead of `mapMaybe (\n -> M.lookup n classMap)`?

Reply:
Great call :) That cut out a lot of code.